Central-Server/API/internal/threads.py

```diff
@@ -0,0 +1,398 @@
+        """Registers callbacks for connection, message, and disconnect events."""
+        @self.socketio_server.event
+        def connect(sid: str, environ):
+            """Websocket connection event, called when a client connects"""
+            on_connect_callback(sid, environ)
+
+        @self.socketio_server.event
+        def my_message(sid: str, data):
+            """Websocket message event, called when a message is sent CLIENT -> SERVER"""
+            on_message_callback(sid, data)
+
+        @self.socketio_server.event
+        def disconnect(sid):
+            """Websocket disconnect event, called when a client disconnects"""
+            on_disconnect_callback(sid)
+
+    def __init__(self, websocket_port): 
+        """Initializes the server on a given port"""
+        threading.Thread.__init__(self) 
+        self.thread_name = "Datastreamer-websocket" 
+        self.thread_id = "M-ws" # Main-websocket (thread_id doesn't matter, as long as it doesn't overlap)
+        self.running = True
+        self.websocket_port = websocket_port
+
+        # Initialize socket communication to Datastreamer
+        self.socketio_server = socketio.Server(cors_allowed_origins='*', async_mode="threading")
+
+        # This line serves ./static/ws_page.html when attempting to connect to the websocket page with http.
+        self.socketio_app = socketio.WSGIApp(self.socketio_server, static_files={
+            '/': {'content_type': 'text/html', 'filename': './static/ws_page.html'}
+        })
+
+        print("(DS-socketio) initializing server..", flush=True)
+
+    def run(self):
+        """This function is called to initialize the thread."""
+        print("(DS-socketio) DS Websocket initialized on " + str(self.websocket_port), flush=True)
+        eventlet.wsgi.server(eventlet.listen(('', self.websocket_port)), self.socketio_app) # Start WSGI server for websocket
+
+
+class BoardThread(threading.Thread): 
+    """Signifies a single datastreamer board connection."""
+    packet_buffer = None # The output buffer of this thread
+
+    def __init__(self, thread_name: str, thread_id: str, board_id: int, communication_channel: communication_interface.CommunicationChannel, pop_back_callback: callable) -> None: 
+        """Generates a datastreamer board thread
+        @thread_name: The name to give to this thread (arbitrary)
+        @thread_id: The ID for this thread (must be unique!)
+        @board_id: The ID of the board to which this thread is connected (must be unique!)
+        @communication_channel: The method that is used to communicate with the Datastreamer
+        @pop_back_callback: The function to be called when this thread needs to pop a job back in the queue."""
+        threading.Thread.__init__(self) 
+        self.thread_name = thread_name 
+        self.communication_channel = communication_channel
+        self.thread_id = thread_id 
+        self.cur_job_config: packets.DataPacket = None
+        self.has_job_config = False

```

Comment:
in theory do we need ```self.has_job_config```,  could we replace with a check for ```self.cur_job_config==None``` in a seperate function